esi/lib/nodes.py

```diff
@@ -138,3 +140,95 @@ def network_list(connection, filter_node=None, filter_network=None):
+    :param mac_address: The mac address of the bare metal port to attach
+
+    :returns: a dictionary with the resulting node and network information
+    {
+        'node': openstack.baremetal.v1.node.Node,
+        'ports': [openstack.network.v2.port.Port]
+        'networks': [openstack.network.v2.network.Network]
+    }
+    """
+
+    network = kwargs.get('network')
+    port = kwargs.get('port')
+    trunk = kwargs.get('trunk')
+    mac_address = kwargs.get('mac_address')
+
+    if (network and port) or (network and trunk) or (port and trunk):
+        raise exceptions.InvalidRequest('Specify only one of network, port or trunk')
+    if not network and not port and not trunk:
+        raise exceptions.InvalidRequest('You must specify either network, port, or trunk')
+
+    if network:
+        parent_network = connection.network.find_network(network, ignore_missing=False)
+        network_port = None
+    elif port:
+        network_port = connection.network.find_port(port, ignore_missing=False)
+    elif trunk:
+        trunk_network = connection.network.find_trunk(trunk, ignore_missing=False)
+        network_port = None
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        f1 = executor.submit(connection.baremetal.get_node, node)
+        f2 = executor.submit(connection.session.get_endpoint, service_type='baremetal', service_name='ironic', interface='public')
+        node = f1.result()
+        baremetal_endpoint = f2.result()
+
+    if mac_address:
+        baremetal_ports = list(connection.baremetal.ports(details=True, address=mac_address))
+        if len(baremetal_ports) == 0:
+            raise exceptions.ResourceFailure('MAC address {0} does not exist on node {1}'.format(mac_address, node.name))
+    else:
+        baremetal_ports = connection.baremetal.ports(details=True, node=node.id)
+        has_free_port = False
+        for bp in baremetal_ports:
+            if 'tenant_vif_port_id' not in bp.internal_info:
+                has_free_port = True
+                break
+
+        if not has_free_port:
+            raise exceptions.ResourceFailure('Node {0} has no free ports'.format(node.name))
+
+    if network:
+        port_name = 'esi-{0}-{1}'.format(node.name, parent_network.name)
+        network_ports = list(connection.network.ports(name=port_name, status='DOWN'))
+        if len(network_ports) > 0:
+            network_port = network_ports[0]
+        else:
+            network_port = connection.network.create_port(name=port_name, network_id=parent_network.id, device_owner='baremetal:none')
+    elif trunk:
+        network_port = connection.network.find_port(trunk_network.port_id, ignore_missing=False)
+
+    data = {'id': network_port.id}
+    if mac_address:
+        data['port_uuid'] = baremetal_ports[0].id
+    connection.session.post('{0}/v1/nodes/{1}/vifs'.format(baremetal_endpoint, node.name), json=data, headers={'X-OpenStack-Ironic-API-Version': '1.69'})
```

Comment:
Is this `connection.node.vif_attach`? Also for API functions it's usually safer to pass in the node's UUID; I think the original code fetches the node from whatever the user passes in.

Reply:
Ah, I see. It's going to be preferable to use the python-ironicclient call here; there's probably a way to instantiate one using the connection info, I'd imagine.

Reply:
Was the purpose of having the connection object being passed in so that no other clients would be created? Or is this an exception?

Reply:
Fair point. I'd say this is an exception, and perhaps there should be a TODO comment mentioning that the logic in the openstacksdk should be updated to allow for the added arguments.

From that perspective, I guess the POST you do is fine! The one thing is that I think you should set 1.69 as a constant somewhere.